lib/stdlib/src/epp.erl

```diff
@@ -981,16 +990,23 @@ scan_include1(_Toks, Inc, From, St) ->
 %%  normal search path, if not we assume that the first directory name
 %%  is a library name, find its true directory and try with that.
 
-expand_lib_dir(Name) ->
+expand_lib_dir(Name, St) ->
     try
-	[App|Path] = filename:split(Name),
-	LibDir = code:lib_dir(list_to_atom(App)),
-	{ok,fname_join([LibDir|Path])}
+        [AppName|Path] = filename:split(Name),
+        App = list_to_atom(AppName),
+        case expand_lib_dir_1(App, St#epp.lib_paths) of
+             LibDir when is_list(LibDir) -> {ok, fname_join([LibDir | Path])};
+             _ -> error
+        end
     catch
-	_:_ ->
-	    error
+        _:_ -> error
     end.
 
+expand_lib_dir_1(App, undefined) ->
+    code:lib_dir(App);
+expand_lib_dir_1(App, Paths) ->
+    proplists:get_value(App, Paths).
+
```

Comment:
ah, that's a neat way to hide it all the way down.

Reply:
I don't really like a solution where the user (or tool) has to specify an individual path per application. This makes it complicated to use and I don't understand what problem it solves compared to just using the include path as it works today (because it works for include lib as well). I did not understand what @ferd had against that solution.

Reply:
The current problem with it is that a system that calls the compiler may do so for more than a single application. For example, you might want to run multiple builds in parallel for various OTP applications.

Today this can only work if all the applications you want to build in parallel have the same paths loaded in memory. If they're unrelated (or if your tool you write to wrap the compiler needs some code present) and that they have shared dependencies of various versions, the files included may clash.

Much like in the bug issue, one example is for us in rebar3, where your own application may include headers from an application that rebar3 (or one of its installed plugins) requires. If both versions aren't the same, the application being compiled may end up with header files from the build tool or its plugins, and cause issues. We wouldn't have a repeatable build.

Currently we work around this by compiling all OTP applications sequentially, and managing global code paths such that when the compiler is loaded, we unload clashing modules and apps (needed for behaviours and parse transforms), change the code paths (needed to load the right modules and for include files), and then call the compiler. Once this is done, we must revert them all.

This new patch would let us at least do something like run concurrent builds as long as we do not detect clashes between include paths and running parse transforms/behaviours, which is a much smaller set of clashing applications, with a much smaller set of manipulations to get right (behaviour modules and parse transform modules are rarely needed at runtime by the build tool, aside for building and compiling). 

For us it is simpler to specify all the paths explicitly than it would be to juggle them live in memory while executing that same code we're trying to unload. So this functional interface is simplifying things for us (including potentially running parallel tests) rather than complexifying them, but I understand that for the regular user, this is not a concern.

Reply:
See https://github.com/erlang/rebar3/blob/master/src/rebar_paths.erl as the new code we are introducing to fix some bugs we had with path handling.